ayeletK/assignment3/grid.cpp

```diff
@@ -0,0 +1,120 @@
+#include "position.cpp"
+#include "dictionary.cpp"
+
+class Grid{
+    int _rows, _cols;
+    vector<vector<char>> _grid;
+    bool update_location(int* row, int* col){
+        if (*col + 1 < _cols){ 
+            (*col)++;
+            return true;
+        }
+        else if (*row + 1 < _rows) {
+            *col = 0;
+            (*row)++;
+            return true;
+        }
+        return false;
+    }
+    bool legal_row(int row){
+        return row >=0 && row < _rows;
+    }
+    bool legal_col(int col){
+        return col >=0 && col < _cols;
+    }
+    bool legal_position(Position pos){
+        return legal_row(pos._row) && legal_col(pos._col);
+    }
+    set<Position> get_neighbours(Position curr_pos){
+        Position neighbour;
+        set<Position> neighbours;
+        for(int row=curr_pos._row - 1; row <= curr_pos._row + 1; ++row){
+            for(int col=curr_pos._col - 1; col <= curr_pos._col + 1; ++col){
+                neighbour = Position(row, col);
+                if (legal_position(neighbour) && !curr_pos.equal(neighbour)) {
+                    neighbours.insert(neighbour);
+                }
+            }
+        }
+        return neighbours;
+    }
+    set<Position> get_unvisited_neighbours(Position curr_pos, set<Position>* visited){
+        set<Position> res;
+        for (Position pos: get_neighbours(curr_pos)){
+            if (visited->find(pos) == visited->end())
+                res.insert(pos);
+        }
+        return res;
+    }
+    void words_with_prefix(Position curr_pos, set<Position>* visited, 
+    Dictionary* dict, vector<char>* prefix, set<string>* words){
+        prefix->push_back(get_value(curr_pos));
+        visited->insert(curr_pos);
+        string curr_prefix = string(prefix->begin(), prefix->end());
+        if (! dict->is_prefix(curr_prefix)) {
+            visited->erase(curr_pos);
+            prefix->pop_back();
+            return;
+        }
+        if (dict->is_word(curr_prefix))
+            words->insert(curr_prefix);
+        for (Position neighbour: get_unvisited_neighbours(curr_pos, visited))
+            words_with_prefix(neighbour, visited, dict, prefix, words);
+        visited->erase(curr_pos);
+        prefix->pop_back();
+    }
+    set<string> find_words_from_position(Position start_pos, Dictionary* dict){
+        set<string> res;
+        vector<char> prefix;
```

Comment:
Like above, `prefix` can be a string. You can still use the same `push_back`,`pop_back` interface.

Reply:
👍